Description: Fix build failures with -Werror=format-security
Author: Kurt Roeckx <kurt@roeckx.be>
Last-Update: 2015-07-20

Index: ntp-4.2.8p3+dfsg/sntp/unity/unity.c
===================================================================
--- ntp-4.2.8p3+dfsg.orig/sntp/unity/unity.c
+++ ntp-4.2.8p3+dfsg/sntp/unity/unity.c
@@ -326,8 +326,7 @@ void UnityConcludeTest(void)
 				printf(" ");
 			}
 
-			printf("| ");
-			printf(Unity.XFAILMessage);
+			printf("| %s", Unity.XFAILMessage);
 			Unity.XFAILMessage = NULL;
 		}
 		else
