Index: ntp-4.2.4p0+dfsg/util/ntp-keygen.c
===================================================================
--- ntp-4.2.4p0+dfsg.orig/util/ntp-keygen.c	2007-03-07 12:02:30.000000000 +0100
+++ ntp-4.2.4p0+dfsg/util/ntp-keygen.c	2007-04-10 11:05:40.000000000 +0200
@@ -165,7 +165,7 @@
 #ifdef OPENSSL
 u_int	modulus = PLEN;		/* prime modulus size (bits) */
 #endif
-int	nkeys = 0;		/* MV keys */
+int	nkeys = 1;		/* MV keys */
 time_t	epoch;			/* Unix epoch (seconds) since 1970 */
 char	*hostname;		/* host name (subject name) */
 char	*trustname;		/* trusted host name (issuer name) */
