Index: ntp-4.2.4p0+dfsg/configure.ac
===================================================================
--- ntp-4.2.4p0+dfsg.orig/configure.ac	2007-04-10 10:37:30.000000000 +0200
+++ ntp-4.2.4p0+dfsg/configure.ac	2007-04-10 11:08:16.000000000 +0200
@@ -443,7 +443,7 @@
 AC_CHECK_HEADERS(netinfo/ni.h, [AC_DEFINE(HAVE_NETINFO, 1, [NetInfo support?])])
 AC_CHECK_HEADERS(sun/audioio.h sys/audioio.h)
 dnl AC_CHECK_HEADERS(sys/chudefs.h)
-AC_CHECK_HEADERS(sys/clkdefs.h sys/file.h)
+AC_CHECK_HEADERS(sys/clkdefs.h sys/file.h sys/var.h)
 case "$host" in
  *-*-sunos4*) ;;
  *) AC_CHECK_HEADERS(sys/ioctl.h)
@@ -530,6 +530,17 @@
     ;;
 esac
 
+AC_CHECK_HEADER(libelf/nlist.h,
+[AC_DEFINE(LIBELF_NLIST_STRUCT, 1, [nlist stuff])
+AC_CACHE_CHECK([for n_un in struct nlist], ac_cv_struct_nlist_n_un,
+[AC_TRY_COMPILE([#include <libelf/nlist.h>],
+[struct nlist n; n.n_un.n_name = 0;],
+ac_cv_struct_nlist_n_un=yes, ac_cv_struct_nlist_n_un=no)])
+if test $ac_cv_struct_nlist_n_un = yes; then
+  AC_DEFINE(NLIST_NAME_UNION, 1, [does struct nlist use a name union?])
+fi
+])dnl
+
 dnl AC_CHECK_HEADERS(net/if.h, [], [],
 dnl [#if HAVE_SYS_TYPES_H
 dnl # include <sys/types.h>
Index: ntp-4.2.4p0+dfsg/util/tickadj.c
===================================================================
--- ntp-4.2.4p0+dfsg.orig/util/tickadj.c	2004-02-25 06:58:33.000000000 +0100
+++ ntp-4.2.4p0+dfsg/util/tickadj.c	2007-04-10 11:08:16.000000000 +0200
@@ -176,11 +176,15 @@
 
 #ifdef NLIST_STRUCT
 # include <nlist.h>
+#elif LIBELF_NLIST_STRUCT
+# include <libelf/nlist.h>
 #else /* not NLIST_STRUCT */ /* was defined(SYS_AUX3) || defined(SYS_AUX2) */
 # include <sys/resource.h>
 # include <sys/file.h>
 # include <a.out.h>
-# include <sys/var.h>
+# ifdef HAVE_SYS_VAR_H
+#  include <sys/var.h>
+# endif
 #endif
 
 #include "ntp_io.h"
